src/Jackalope/Version/VersionHandler.php

```diff
@@ -119,27 +130,50 @@ public function checkinItem($path)
 
         // TODO set subgraph to read only
 
-        $versionHistoryNode = $node->getPropertyValue('jcr:versionHistory');
+        /** @var NodeInterface $versionHistoryNode */
+        $versionHistoryUuid = $node->getProperty('jcr:versionHistory')->getString();
+        $versionHistoryNode = $this->objectManager->getNodeByIdentifier($versionHistoryUuid, 'Version\VersionHistory');
 
         // FIXME add some kind of sharding
         // should avoid to have too many nodes on the same level
         $versionNode = $versionHistoryNode->addNode(UUIDHelper::generateUUID(), 'nt:version');
         $versionNode->setProperty('jcr:uuid', UUIDHelper::generateUUID());
         $versionNode->setProperty('jcr:created', new \DateTime());
-
-        // TODO create frozen node
-
-        $versionNode->setProperty('jcr:predecessors', $node->getProperty('jcr:predecessors')->getString());
-        $node->setProperty('jcr:predecessors', array(), PropertyType::REFERENCE, false);
-        foreach ($versionNode->getProperty('jcr:predecessors') as $predecessorUuid) {
-            $predecessor = $this->objectManager->getNodeByIdentifier($predecessorUuid);
-            $predecessor->setProperty('jcr:successors', array_merge($predecessor->getPropertyValueWithDefault('jcr:successors', array()), array($versionNode)));
+        $versionNode->setProperty('jcr:successors', array());
+
+        $this->createFrozenNode($versionNode, $node);
+
+        $baseVersionNode->setProperty('jcr:successors', array($versionNode), PropertyType::REFERENCE, false);
+        $baseVersionNode->setModified();
+        $versionNode->setProperty(
+            'jcr:predecessors',
+            $node->getProperty('jcr:predecessors')->getString(),
+            PropertyType::REFERENCE
+        );
+        $versionNode->setProperty(
+            'jcr:successors',
+            array(),
+            PropertyType::REFERENCE,
+            false
+        );
+
+        $node->setProperty(
+            'jcr:predecessors',
+            array($versionNode),
+            PropertyType::REFERENCE,
+            false
+        ); // TODO double check with specification, this line seems not to be part of it
```

Comment:
It's not part of the specification, but definitely makes sense. The spec says that you should copy the predecessors from the actual node, which will never have a predecessor without this statement. So i guess I just remove the comment, if this is ok for you.